Some unit tests are failing in `firestarter-workflows` application

Python 3.11.5
poetry 1.8.3

```
poetry install
poetry install -E build_images
```

```
$ poetry run pytest                                               
==================================================================================================================================================================================================================================== test session starts =====================================================================================================================================================================================================================================
platform linux -- Python 3.11.5, pytest-7.4.4, pluggy-1.5.0
rootdir: /tmp/firestarter-workflows
plugins: anyio-3.6.2
collected 7 items                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

firestarter/tests/test_dagger_workflow.py F                                                                                                                                                                                                                                                                                                                                                                                                                                            [ 14%]
firestarter/tests/test_preprocessor.py ...                                                                                                                                                                                                                                                                                                                                                                                                                                             [ 57%]
firestarter/tests/test_workflow_validate.py ...                                                                                                                                                                                                                                                                                                                                                                                                                                        [100%]

========================================================================================================================================================================================================================================== FAILURES ==========================================================================================================================================================================================================================================
____________________________________________________________________________________________________________________________________________________________________________________________________________________________________ test_dagger_workflow ____________________________________________________________________________________________________________________________________________________________________________________________________________________________________

    def test_dagger_workflow() -> None:
    
        def myTestFunc(context: Context) -> None:
            tm: TaskManager = TaskManager()
            tm.context = context
            tm.load(
                os.path.join(
                    os.path.dirname(__file__), 'fixtures/test_workflow.yaml',
                ),
                '../workflows/ci/schema.json',
            )
            tm.run()
    
>       anyio.run(lambda: Context().start(myTestFunc))

firestarter/tests/test_dagger_workflow.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/alambike/.cache/pypoetry/virtualenvs/firestarter-workflows-guuNLrIx-py3.11/lib/python3.11/site-packages/anyio/_core/_eventloop.py:70: in run
    return asynclib.run(func, *args, **backend_options)
/home/alambike/.cache/pypoetry/virtualenvs/firestarter-workflows-guuNLrIx-py3.11/lib/python3.11/site-packages/anyio/_backends/_asyncio.py:292: in run
    return native_run(wrapper(), debug=debug)
/home/alambike/.asdf/installs/ivm-python/3.11.5/lib/python3.11/asyncio/runners.py:190: in run
    return runner.run(main)
/home/alambike/.asdf/installs/ivm-python/3.11.5/lib/python3.11/asyncio/runners.py:118: in run
    return self._loop.run_until_complete(task)
/home/alambike/.asdf/installs/ivm-python/3.11.5/lib/python3.11/asyncio/base_events.py:653: in run_until_complete
    return future.result()
/home/alambike/.cache/pypoetry/virtualenvs/firestarter-workflows-guuNLrIx-py3.11/lib/python3.11/site-packages/anyio/_backends/_asyncio.py:287: in wrapper
    return await func(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <firestarter.workflows.ci.dagger_context.Context object at 0x778bcde03f50>, fn = <function test_dagger_workflow.<locals>.myTestFunc at 0x778bcddf9ee0>

    async def start(self, fn: Callable) -> None:
>       with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
E       TypeError: 'Connection' object does not support the context manager protocol

firestarter/workflows/ci/dagger_context.py:18: TypeError
================================================================================================================================================================================================================================== short test summary info ===================================================================================================================================================================================================================================
FAILED firestarter/tests/test_dagger_workflow.py::test_dagger_workflow - TypeError: 'Connection' object does not support the context manager protocol
================================================================================================================================================================================================================================ 1 failed, 6 passed in 0.66s =====
```

--- a/.github/workflows/firestarter_pr-verify.yaml
+++ b/.github/workflows/firestarter_pr-verify.yaml
@@ -15,6 +15,9 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      - name: Install poetry
+        run: pipx install poetry
+
       - name: Set up Python 3.11
         uses: actions/setup-python@v5
         with:

--- a/.github/workflows/firestarter_pr-verify.yaml
+++ b/.github/workflows/firestarter_pr-verify.yaml
@@ -1,0 +1,31 @@
+name: Verify firestarter workflows python package
+
+on:
+  pull_request:
+    paths:
+      - 'firestarter/**'
+    types:
+      - opened
+      - synchronize
+      - reopened
+
+jobs:
+  unit-tests-firestarter-python-package:
+    runs-on: ubuntu-24.04
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v5
+        with:
+          python-version: 3.11
+          cache: 'poetry'
+
+      - name: Install dependencies
+        run: |
+          poetry install --all-extras
+
+      - name: Run tests
+        run: |
+          poetry run pytest
+

--- a/.github/workflows/firestarter_pr-verify_fallback.yaml
+++ b/.github/workflows/firestarter_pr-verify_fallback.yaml
@@ -2,8 +2,6 @@ name: Verify firestarter workflows python package
 
 on:
   pull_request:
-    paths:
-      - 'firestarter/**'
     types:
       - opened
       - synchronize

--- a/.github/workflows/firestarter_pr-verify_fallback.yaml
+++ b/.github/workflows/firestarter_pr-verify_fallback.yaml
@@ -1,0 +1,19 @@
+name: Verify firestarter workflows python package
+
+on:
+  pull_request:
+    paths:
+      - 'firestarter/**'
+    types:
+      - opened
+      - synchronize
+      - reopened
+
+jobs:
+  unit-tests-firestarter-python-package:
+    runs-on: ubuntu-24.04
+    if: 'false'
+    name: Code / Verify
+    steps:
+      - run: 'echo "No Code / Verify required"'
+

--- a/firestarter/tests/test_dagger_workflow.py
+++ b/firestarter/tests/test_dagger_workflow.py
@@ -6,7 +6,7 @@ import os
 
 def test_dagger_workflow() -> None:
 
-    def myTestFunc(context: Context) -> None:
+    async def myTestFunc(context: Context) -> None:
         tm: TaskManager = TaskManager()
         tm.context = context
         tm.load(
@@ -15,7 +15,7 @@ def test_dagger_workflow() -> None:
             ),
             '../workflows/ci/schema.json',
         )
-        tm.run()
+        await tm.run()
 
     anyio.run(lambda: Context().start(myTestFunc))
 

--- a/firestarter/workflows/ci/bash_runner.py
+++ b/firestarter/workflows/ci/bash_runner.py
@@ -66,7 +66,7 @@ def exec_run_in_container(commands: List[str], container, dagger_client):
         # Mount the script directory
         .with_mounted_directory(temp_dir, src)
         # Execute the bash script
-        .withExec([f"{file_name}"])
+        .with_exec([f"{file_name}"])
     )
 
     return container

--- a/firestarter/workflows/ci/bash_runner.py
+++ b/firestarter/workflows/ci/bash_runner.py
@@ -22,7 +22,7 @@ def create_script_from_commands(commands: List[str], temp_dir: str) -> None:
 
     return temp_file.name
 
-def exec_run_in_container(commands: List[str], container, dagger_client):
+async def exec_run_in_container(commands: List[str], container, dagger_client):
     """
     Executes commands in a container
 
@@ -43,7 +43,7 @@ def exec_run_in_container(commands: List[str], container, dagger_client):
         .with_entrypoint(["bash", "-e"])
     )
 
-    if container.workdir() != WORKSPACE_PATH:
+    if await container.workdir() != WORKSPACE_PATH:
         container = (
             container
             # Mount current dir
@@ -65,6 +65,8 @@ def exec_run_in_container(commands: List[str], container, dagger_client):
         container
         # Mount the script directory
         .with_mounted_directory(temp_dir, src)
+        # Make bash script executable
+        .with_exec(["chmod", "+x", f"{file_name}"])
         # Execute the bash script
         .with_exec([f"{file_name}"])
     )

--- a/firestarter/workflows/ci/dagger_context.py
+++ b/firestarter/workflows/ci/dagger_context.py
@@ -15,7 +15,7 @@ class Context:
         self.secrets: dict = secrets if secrets else {}
 
     async def start(self, fn: Callable) -> None:
-        with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
+        async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
             self.dagger_client = client
             fn(self)
 

--- a/firestarter/workflows/ci/dagger_context.py
+++ b/firestarter/workflows/ci/dagger_context.py
@@ -17,7 +17,7 @@ class Context:
     async def start(self, fn: Callable) -> None:
         async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
             self.dagger_client = client
-            fn(self)
+            await fn(self)
 
     @property
     def dagger_client(self) -> dagger.Client:
@@ -57,6 +57,6 @@ class Context:
     def new_container(self, image: str) -> dagger.Container:
         return self.dagger_client.container().from_(image)
 
-    def set_output(self, name: str, container: dagger.Container) -> None:
+    async def set_output(self, name: str, container: dagger.Container) -> None:
         self.container = container
-        self.outputs[name] = container.stdout()
+        self.outputs[name] = await container.stdout()

--- a/firestarter/workflows/ci/task_manager.py
+++ b/firestarter/workflows/ci/task_manager.py
@@ -32,6 +32,6 @@ class TaskManager:
         # let's load the tasks
         task_scheduler(config_data, self.tasks)
 
-    def run(self) -> None:
-        self.tasks.run_tasks(self.context)
+    async def run(self) -> None:
+        await self.tasks.run_tasks(self.context)
 

--- a/firestarter/workflows/ci/tasks.py
+++ b/firestarter/workflows/ci/tasks.py
@@ -10,12 +10,13 @@ class Task:
         self.env: dict = args.get("vars", {})
 
     def execute(self, ctx) -> None:
-        container = self.prepare_ctx(ctx)
-        container = exec_run_in_container(
-            self.commands, container, ctx.dagger_client
-        )
+        try:
+            container = self.prepare_ctx(ctx)
+            container = exec_run_in_container(
+                self.commands, container, ctx.dagger_client
+            )
 
-        if container.exit_code() != 0:
+        except Exception:
             raise f"ERROR: {self.name}: {container.stderr()} {container.stdout()}"
 
         ctx.set_output(self.name, container)

--- a/firestarter/workflows/ci/tasks.py
+++ b/firestarter/workflows/ci/tasks.py
@@ -9,17 +9,17 @@ class Task:
         self.image: str = args.get("image", None)
         self.env: dict = args.get("vars", {})
 
-    def execute(self, ctx) -> None:
+    async def execute(self, ctx) -> None:
         try:
             container = self.prepare_ctx(ctx)
-            container = exec_run_in_container(
+            container = await exec_run_in_container(
                 self.commands, container, ctx.dagger_client
             )
 
         except Exception:
-            raise f"ERROR: {self.name}: {container.stderr()} {container.stdout()}"
+            raise f"ERROR: {self.name}: {await container.stderr()} {await container.stdout()}"
 
-        ctx.set_output(self.name, container)
+        await ctx.set_output(self.name, container)
 
     def prepare_ctx(self, ctx):
         container = ctx.next_container(self.image)
@@ -48,8 +48,8 @@ class TaskGroup:
 
         self.__tasks[task.name] = task
 
-    def run_tasks(self, context) -> None:
+    async def run_tasks(self, context) -> None:
         ctx = context
 
         for task in sorted(self.__tasks.values(), key=lambda t: t.order):
-            task.execute(ctx)
+            await task.execute(ctx)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -27,6 +27,9 @@ build_images = ["docker", "azure-identity", "azure-keyvault-secrets", "boto3"]
 [tool.poetry.scripts]
 firestarter = "firestarter.cli:main"
 
+[tool.pytest.ini_options]
+pythonpath = ["."]
+
 [build-system]
 requires = ["poetry-core"]
 build-backend = "poetry.core.masonry.api"

